satellite/satellitedb/gracefulexit_test.go

```diff
@@ -0,0 +1,151 @@
+
+		testData := []struct {
+			nodeID      storj.NodeID
+			bytes       int64
+			transferred int64
+			failed      int64
+		}{
+			{testrand.NodeID(), 10, 2, 1},
+			{testrand.NodeID(), 1, 4, 0},
+		}
+		for _, data := range testData {
+			err := geDB.IncrementProgress(ctx, data.nodeID, data.bytes, data.transferred, data.failed)
+			require.NoError(t, err)
+
+			progress, err := geDB.GetProgress(ctx, data.nodeID)
+			require.NoError(t, err)
+			require.Equal(t, data.bytes, progress.BytesTransferred)
+			require.Equal(t, data.transferred, progress.PiecesTransferred)
+			require.Equal(t, data.failed, progress.PiecesFailed)
+
+			err = geDB.IncrementProgress(ctx, data.nodeID, 1, 1, 1)
+			require.NoError(t, err)
+
+			progress, err = geDB.GetProgress(ctx, data.nodeID)
+			require.NoError(t, err)
+			require.Equal(t, data.bytes+1, progress.BytesTransferred)
+			require.Equal(t, data.transferred+1, progress.PiecesTransferred)
+			require.Equal(t, data.failed+1, progress.PiecesFailed)
+		}
+	})
+}
+
+func TestTransferQueueItem(t *testing.T) {
+	// test basic graceful exit transfer queue crud
+	satellitedbtest.Run(t, func(t *testing.T, db satellite.DB) {
+		ctx := context.Background()
+
+		geDB := db.GracefulExit()
+
+		nodeID1 := testrand.NodeID()
+		nodeID2 := testrand.NodeID()
+		path1 := testrand.Bytes(memory.B * 32)
+		path2 := testrand.Bytes(memory.B * 32)
+		items := []gracefulexit.TransferQueueItem{
+			{
+				NodeID:          nodeID1,
+				Path:            path1,
+				PieceNum:        1,
+				DurabilityRatio: 0.9,
+			},
+			{
+				NodeID:          nodeID1,
+				Path:            path2,
+				PieceNum:        2,
+				DurabilityRatio: 1.1,
+			},
+			{
+				NodeID:          nodeID2,
+				Path:            path1,
+				PieceNum:        2,
+				DurabilityRatio: 0.9,
+			},
+			{
+				NodeID:          nodeID2,
+				Path:            path2,
+				PieceNum:        1,
+				DurabilityRatio: 1.1,
+			},
+		}
+
+		// test basic create, update, get delete
+		err := geDB.Enqueue(ctx, items)
+		require.NoError(t, err)
+
+		for _, tqi := range items {
+			item, err := geDB.GetTransferQueueItem(ctx, tqi.NodeID, tqi.Path)
+			require.NoError(t, err)
+
+			item.DurabilityRatio = 1.2
+			item.RequestedAt = time.Now()
+
+			err = geDB.UpdateTransferQueueItem(ctx, *item)
+			require.NoError(t, err)
+
+			latestItem, err := geDB.GetTransferQueueItem(ctx, tqi.NodeID, tqi.Path)
+			require.NoError(t, err)
+			require.Equal(t, item.DurabilityRatio, latestItem.DurabilityRatio)
+			require.True(t, item.RequestedAt.Truncate(time.Millisecond).Equal(latestItem.RequestedAt.Truncate(time.Millisecond)))
+		}
+
+		// mark the first item finished
+		item, err := geDB.GetTransferQueueItem(ctx, nodeID1, path1)
+		require.NoError(t, err)
+
+		item.FinishedAt = time.Now()
+		err = geDB.UpdateTransferQueueItem(ctx, *item)
+		require.NoError(t, err)
+
+		queueItems, err := geDB.GetIncomplete(ctx, nodeID1, 10, 0)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(queueItems))
```

Comment:
nit - `require.Len(t, queueItems, 1)` should also work

Reply:
It might also be worth testing `GetIncomplete` before updating with `FinishedAt`, and making sure the length of the items returned is 2.

Reply:
Updated